File: apps/web/src/lib/log.ts
/**
 * One structured JSON line per meaningful event. Never pass names, emails, tokens, audio or
 * transcript text in `fields`: ids only. Drill events always carry `drillId`.
 */

type Level = 'info' | 'warn' | 'error';
type Fields = Record<string, string | number | boolean | null | undefined>;

function write(level: Level, event: string, fields: Fields = {}): void {
  const line = JSON.stringify({ ts: new Date().toISOString(), level, event, ...fields });
  (level === 'info' ? process.stdout : process.stderr).write(`${line}\n`);
}

export const log = {
  info: (event: string, fields?: Fields) => write('info', event, fields),
  warn: (event: string, fields?: Fields) => write('warn', event, fields),
  error: (event: string, fields?: Fields) => write('error', event, fields),
};

/** Error name and message only. Stack traces and SDK payloads stay out of the log line. */
export function errorFields(error: unknown): Fields {
  if (error instanceof Error) return { errorName: error.name, errorMessage: error.message.slice(0, 300) };
  return { errorName: 'unknown' };
}

File: apps/agent/chaukanna_agent/log.py
"""One structured JSON line per meaningful event, always carrying drillId. Ids and labels only:
never transcript text, quotes, or audio."""

from __future__ import annotations

import json
import logging
import sys

from .clock import utc_now_iso

_events = logging.getLogger("chaukanna_agent.events")


def event(name: str, drill_id: str, level: int = logging.INFO, **fields: object) -> None:
    line = {"ts": utc_now_iso(), "event": name, "drillId": drill_id, **fields}
    _events.log(level, json.dumps(line, ensure_ascii=False, default=str))


def configure_logging(level: int = logging.INFO) -> None:
    """JSON event lines to stderr. The SDK logs transcript previews at DEBUG, so it stays at WARNING."""
    handler = logging.StreamHandler(sys.stderr)
    handler.setFormatter(logging.Formatter("%(message)s"))
    _events.handlers[:] = [handler]
    _events.setLevel(level)
    _events.propagate = False
    logging.getLogger("strands").setLevel(logging.WARNING)
